src/Conversion/ONNXToMhlo/NN/Conv.cpp

```diff
@@ -0,0 +1,147 @@
+//
+// Copyright 2019-2022 The IBM Research Authors.
+//
+// =============================================================================
+//
+// This file lowers ONNX Conv Operators to Mhlo dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "src/Conversion/ONNXToMhlo/ONNXToMhloCommon.hpp"
+#include "src/Dialect/ONNX/ShapeInference/ONNXShapeHelper.hpp"
+#include "src/Support/TypeUtilities.hpp"
+
+using namespace mlir;
+
+namespace onnx_mlir {
+
+namespace {
+
+struct ONNXConvOpLoweringToMhlo : public ConversionPattern {
+  ONNXConvOpLoweringToMhlo(MLIRContext *ctx)
+      : ConversionPattern(mlir::ONNXConvOp::getOperationName(), 1, ctx) {}
+
+  LogicalResult matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const final {
+
+    ONNXConvOpAdaptor operandAdaptor(operands);
+    ONNXConvOp convOp = llvm::dyn_cast<ONNXConvOp>(op);
+    Location loc = op->getLoc();
+
+    ONNXConvOpShapeHelper shapeHelper(&convOp);
+    LogicalResult shapecomputed = shapeHelper.computeShape(operandAdaptor);
+    assert(succeeded(shapecomputed) && "Could not compute output shape");
+
+    llvm::SmallVector<IndexExpr, 2> kernelShape = shapeHelper.kernelShape;
+    llvm::SmallVector<int64_t, 2> strides = shapeHelper.strides;
+    llvm::SmallVector<int64_t, 2> dilations = shapeHelper.dilations;
+    DimsExpr outputDims = shapeHelper.dimsForOutput();
+    int outputRank = shapeHelper.dimsForOutput().size();
+
+    Value inputOperand = operandAdaptor.X();
+    Value filterOperand = operandAdaptor.W();
+    Value biasOperand = operandAdaptor.B();
+    bool hasBias = !biasOperand.getType().isa<NoneType>();
+    int64_t groupNum = convOp.group();
+
+    assert(isRankedShapedType(inputOperand.getType()) &&
+           "Expected Ranked ShapedType");
+    ShapedType inputType = inputOperand.getType().cast<ShapedType>();
+    llvm::ArrayRef<int64_t> inputShape = inputType.getShape();
+    Type outputType = *op->result_type_begin();
+    // Onnx Input is NCHW
+    int64_t spatialOffset = 2;
+    int64_t rank = inputType.getRank();
+    int64_t kernelSize = kernelShape.size();
+
+    SmallVector<int64_t> inputSpatialDimensions;
+    for (int64_t i = spatialOffset; i < rank; i++) {
+      inputSpatialDimensions.push_back(i);
+    }
+
+    SmallVector<int64_t> kernelDimensions;
+    for (int64_t i = spatialOffset; i < spatialOffset + kernelSize; i++) {
+      kernelDimensions.push_back(i);
+    }
+
+    SmallVector<int64_t> outputSpatialDimensions;
+    for (int64_t i = spatialOffset; i < outputRank; i++) {
+      outputSpatialDimensions.push_back(i);
+    }
+
+    // paddings
+    DimsExpr pads = shapeHelper.pads;
+    llvm::StringRef padding = convOp.auto_pad();
+    int64_t spatialRank = rank - spatialOffset;
+    SmallVector<int64_t> flattenPaddings;
+    bool needPadding = (padding == "NOTSET");
+
+    // currently only support static shape
```

Comment:
Yes, we can use the function to test if all index expressions are literals, and we need to modify IndexExpr so we can reuse the result from the shape helper.

Reply:
You could refer to the case below where we need to calculate the padding again. For simplicity we previously just calculate all paddings instead of reuse those in the shape helper.